File: solutions/typescript/dnd-character/2/dnd-character.ts
export class DnDCharacter {
    public readonly strength: number;
    public readonly dexterity: number;
    public readonly constitution: number;
    public readonly intelligence: number;
    public readonly wisdom: number;
    public readonly charisma: number;
    public readonly hitpoints: number;

    constructor() {
        this.strength = DnDCharacter.generateAbilityScore();
        this.dexterity = DnDCharacter.generateAbilityScore();
        this.constitution = DnDCharacter.generateAbilityScore();
        this.intelligence = DnDCharacter.generateAbilityScore();
        this.wisdom = DnDCharacter.generateAbilityScore();
        this.charisma = DnDCharacter.generateAbilityScore();

        this.hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution);
    }

    public static generateAbilityScore(): number {
        const rolls = Array.from(
            { length: 4 },
            () => Math.floor(Math.random() * 6) + 1,
        );

        return rolls
            .sort()
            .slice(-3)
            .reduce((sum, roll) => sum + roll, 0);
    }

    public static getModifierFor(abilityValue: number): number {
        return Math.floor((abilityValue - 10) / 2);
    }
}

File: solutions/c/dnd-character/5/dnd_character.c
#include "dnd_character.h"

#include <math.h>
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>

int ability(void) {
  // Seed random number generator once
  static bool seed_initialized = false;
  if (!seed_initialized) {
    srand((unsigned int)time(NULL));
    seed_initialized = true;
  }

  // Roll 4d6, drop lowest - single pass approach
  int sum = 0;
  int min_roll = 7;

  for (int i = 0; i < 4; i++) {
    int roll = (rand() % 6) + 1;
    sum += roll;
    if (roll < min_roll) min_roll = roll;
  }

  return sum - min_roll;
}

int modifier(int score) { return (int)floor((score - 10) / 2.0); }

dnd_character_t make_dnd_character(void) {
  dnd_character_t character = {.strength = ability(),
                               .dexterity = ability(),
                               .constitution = ability(),
                               .intelligence = ability(),
                               .wisdom = ability(),
                               .charisma = ability()};

  character.hitpoints = 10 + modifier(character.constitution);
  return character;
}
